snippets/SharedHandle.hpp

```diff
@@ -221,49 +122,54 @@ class SharedHandleBase
     return &m_handle;
   }
 
-  size_t add_ref() VULKAN_HPP_NOEXCEPT
+  void reset() VULKAN_HPP_NOEXCEPT
   {
     if ( !m_handle )
-      return 0;
-    return m_control.add_ref();
+      return;
+
+    auto refs = m_control.use_count();
```

Comment:
This is dangerous!
I think, you're still mixing ref-counting and parent referencing.
Maybe, with something like
`using SharedHandle = std::shared_ptr<ShareableHandle>;`
where `ShareableHandle` would hold the actual handle, a deleter functor for that handle, SharedHandles of all the required parents and pools, and maybe more data required for the deleter, reference counting and parent/hierarchy handling could be completely decoupled.

Reply:
never

Reply:
`using SharedHandle = std::shared_ptr<ShareableHandle>;` this is just a complete miss. That is a pure overhead, which doesn't do anything but allocate memory. Handles should be on top of the stack! Else this mechanism is inefficient. 

std::shared_pointer should reference count the handle, not the parent, parent is a shared handle itself, a recursive implementation allows the tree of inheritance to be shallow and a copy should only reference the node it copies.

Reply:
I think I might make critical section with spinlock around the destruction, for it to be safer